src/main/java/dev/openfeature/sdk/FlagMetadata.java

```diff
@@ -0,0 +1,170 @@
+package dev.openfeature.sdk;
+
+import dev.openfeature.sdk.exceptions.GeneralError;
+import dev.openfeature.sdk.exceptions.ParseError;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Immutable Flag Metadata representation. Implementation is backed by a {@link Map} and immutability is provided
+ * through builder and accessors.
+ */
+public class FlagMetadata {
+    private final Map<String, Object> metadata;
+
+    private FlagMetadata(Map<String, Object> metadata) {
+        this.metadata = metadata;
+    }
+
+    /**
+     * Retrieve a {@link String} value for the given key. If a value is not found, {@link GeneralError} will be thrown.
+     * If value exist but of another type, {@link ParseError} will be thrown.
+     *
+     * @param key flag metadata key to retrieve
+     */
+    public String getString(final String key) {
+        return getValue(key, String.class);
+    }
+
+    /**
+     * Retrieve an {@link Integer} value for the given key.
+     * If a value is not found, {@link GeneralError} will be thrown.
+     * If value exist but of another type, {@link ParseError} will be thrown.
+     *
+     * @param key flag metadata key to retrieve
+     */
+    public Integer getInteger(final String key) {
+        return getValue(key, Integer.class);
+    }
+
+    /**
+     * Retrieve an {@link Float} value for the given key. If a value is not found, {@link GeneralError} will be thrown.
+     * If value exist but of another type, {@link ParseError} will be thrown.
+     *
+     * @param key flag metadata key to retrieve
+     */
+    public Float getFloat(final String key) {
+        return getValue(key, Float.class);
+    }
+
+    /**
+     * Retrieve an {@link Double} value for the given key.
+     * If a value is not found, {@link GeneralError} will be thrown.
+     * If value exist but of another type, {@link ParseError} will be thrown.
+     *
+     * @param key flag metadata key to retrieve
+     */
+    public Double getDouble(final String key) {
+        return getValue(key, Double.class);
+    }
+
+    /**
+     * Retrieve an {@link Boolean} value for the given key.
+     * If a value is not found, {@link GeneralError} will be thrown.
+     * If value exist but of another type, {@link ParseError} will be thrown.
+     *
+     * @param key flag metadata key to retrieve
+     */
+    public Boolean getBoolean(final String key) {
+        return getValue(key, Boolean.class);
+    }
+
+    private <T> T getValue(final String key, final Class<T> type) {
+        final Object o = metadata.get(key);
+
+        if (o == null) {
+            throw new GeneralError("key " + key + " does not exist in metadata");
```

Comment:
I _think_ we want to return null here, not throw. That's consistent with the context and structure abstractions in the SDK as well.

I could be wrong about this, open to hear why you think throwing is better.

Reply:
I went with the exception to match with the go-sdk implemenation [1]. I can go ahead with null if we agree to that behavior 

[1] - https://github.com/open-feature/go-sdk/blob/main/pkg/openfeature/client.go#L179

Reply:
I am not familiar with the control- and data-flow of the Java SDK, but I am also in favor of avoiding exceptions. The SDK should never fail at runtime IMHO. If exception are the best way to communicate to the caller, then I suggest explicitly marking the methods with JavaDoc annotations for the exception and having the `throws` annotation in the method signature

Reply:
> I went with the exception to match with the go-sdk implemenation [1]. I can go ahead with null if we agree to that behavior

Ya, I think the thing is that practically, returning an error is fundamentally different than exceptions. I can choose to check the error field or not - I _must_ handle exceptions (unless I like crashes a lot).

Reply:
@thisthat @toddbaert  updated the implementation to avoid exceptions. Now it log(debug level) and return null value for both scenarios